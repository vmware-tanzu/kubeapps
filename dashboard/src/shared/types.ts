// Copyright 2018-2022 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

import { JSONSchemaType } from "ajv";
import { RouterState } from "connected-react-router";
import {
  AvailablePackageDetail,
  AvailablePackageSummary,
  GetAvailablePackageSummariesResponse,
  InstalledPackageDetail,
  InstalledPackageSummary,
  PackageAppVersion,
  ResourceRef,
} from "gen/kubeappsapis/core/packages/v1alpha1/packages";
import {
  DockerCredentials,
  OpaqueCredentials,
  PackageRepositoryAuth_PackageRepositoryAuthType,
  SshCredentials,
  TlsCertKey,
  UsernamePassword,
} from "gen/kubeappsapis/core/packages/v1alpha1/repositories";
import { Plugin } from "gen/kubeappsapis/core/plugins/v1alpha1/plugins";
import { HelmPackageRepositoryCustomDetail } from "gen/kubeappsapis/plugins/helm/packages/v1alpha1/helm";
import { KappControllerPackageRepositoryCustomDetail } from "gen/kubeappsapis/plugins/kapp_controller/packages/v1alpha1/kapp_controller";
import { IOperatorsState } from "reducers/operators";
import { Subscription } from "rxjs";
import { IAuthState } from "../reducers/auth";
import { IClustersState } from "../reducers/cluster";
import { IConfigState } from "../reducers/config";
import { IPackageRepositoryState } from "../reducers/repos";
import { RpcError } from "./RpcError";

export class CustomError extends Error {
  public causes: Error[] | undefined;
  // The constructor is defined so we can later on compare the returned object
  // via err.constructor  == FOO
  constructor(message?: string, causes?: Error[]) {
    super(message);
    Object.setPrototypeOf(this, new.target.prototype);
    this.causes = causes;
    this.checkCauses();
  }
  // Workaround used until RPC code (unary) throws a custom rpc error
  // Check if any RPC error is among the causes
  private checkCauses() {
    if (!this.causes) return;
    for (let i = 0; i < this.causes.length; i++) {
      const cause = this.causes[i];
      if (RpcError.isRpcError(cause)) {
        this.causes[i] = new RpcError(cause);
      }
    }
  }
}

// For 4XX HTTP-alike errors
export class ClientNetworkError extends CustomError {}
// 400
export class BadRequestNetworkError extends ClientNetworkError {}
// 401
export class UnauthorizedNetworkError extends ClientNetworkError {}
// 403
export class ForbiddenNetworkError extends ClientNetworkError {}
// 404
export class NotFoundNetworkError extends ClientNetworkError {}
// 408
export class RequestTimeoutNetworkError extends ClientNetworkError {}
// 409
export class ConflictNetworkError extends ClientNetworkError {}
// 422
export class UnprocessableEntityError extends ClientNetworkError {}
// 429
export class TooManyRequestsNetworkError extends ClientNetworkError {}

// For 5XX HTTP-alike errors
export class ServerNetworkError extends CustomError {}
// 500
export class InternalServerNetworkError extends ServerNetworkError {}
// 501
export class NotImplementedNetworkError extends ServerNetworkError {}
// 503
export class ServerUnavailableNetworkError extends ServerNetworkError {}
// 504
export class GatewayTimeoutNetworkError extends ServerNetworkError {}

// Application-level errors
export class FetchError extends CustomError {}

export class FetchWarning extends CustomError {}

export class CreateError extends CustomError {}

export class UpgradeError extends CustomError {}

export class RollbackError extends CustomError {}

export class DeleteError extends CustomError {}

export type DeploymentEvent = "install" | "upgrade";

export interface IReceivePackagesActionPayload {
  response: GetAvailablePackageSummariesResponse;
  paginationToken: string;
}

export interface IPackageState {
  isFetching: boolean;
  hasFinishedFetching: boolean;
  nextPageToken: string;
  selected: {
    error?: FetchError | Error;
    availablePackageDetail?: AvailablePackageDetail;
    pkgVersion?: string;
    appVersion?: string;
    versions: PackageAppVersion[];
    readme?: string;
    readmeError?: string;
    values?: string;
    schema?: JSONSchemaType<any>;
  };
  items: AvailablePackageSummary[];
  categories: string[];
  size: number;
}

export interface IServiceSpec {
  ports: IPort[];
  clusterIP: string;
  type: string;
}

export interface IServiceStatus {
  loadBalancer: {
    ingress?: Array<{ ip?: string; hostname?: string }>;
  };
}

export interface IPort {
  name: string;
  port: number;
  protocol: string;
  targetPort: string;
  nodePort: string;
}

export interface IHTTPIngressPath {
  path: string;
}

export interface IIngressHTTP {
  paths: IHTTPIngressPath[];
}

export interface IIngressRule {
  host: string;
  http: IIngressHTTP;
}

export interface IIngressTLS {
  hosts: string[];
}

export interface IIngressSpec {
  rules: IIngressRule[];
  tls?: IIngressTLS[];
  backend?: {
    serviceName: string;
    servicePort: number;
  };
}

export interface IResourceMetadata {
  name: string;
  namespace: string;
  annotations: { [key: string]: string };
  ownerReferences?: Array<{
    apiVersion: string;
    blockOwnerDeletion: string;
    kind: string;
    name: string;
    uid: string;
  }>;
  creationTimestamp: string;
  selfLink: string;
  resourceVersion: string;
  deletionTimestamp?: string;
  uid: string;
}

export interface IResource {
  apiVersion: string;
  kind: string;
  type: string;
  spec: any;
  status: any;
  metadata: IResourceMetadata;
}

export interface ISecret {
  apiVersion: string;
  kind: string;
  type: string;
  data: { [s: string]: string };
  metadata: IResourceMetadata;
}

export interface IPackageManifestChannel {
  name: string;
  currentCSV: string;
  currentCSVDesc: {
    annotations: {
      "alm-examples": string;
      capabilities: string;
      categories: string;
      certified: string;
      containerImage: string;
      createdAt: string;
      description: string;
      repository: string;
      support: string;
    };
    description: string;
    displayName: string;
    provider: {
      name: string;
    };
    version: string;
    installModes: [
      { supported: boolean; type: "OwnNamespace" },
      { supported: boolean; type: "SingleNamespace" },
      { supported: boolean; type: "MultiNamespace" },
      { supported: boolean; type: "AllNamespaces" },
    ];
    customresourcedefinitions: {
      owned: Array<{
        description: string;
        displayName: string;
        kind: string;
        name: string;
        version: string;
      }>;
    };
  };
}

export interface IPackageManifestStatus {
  catalogSource: string;
  catalogSourceDisplayName: string;
  catalogSourceNamespace: string;
  catalogSourcePublisher: string;
  provider: {
    name: string;
  };
  defaultChannel: string;
  channels: IPackageManifestChannel[];
}

export interface IPackageManifest extends IResource {
  status: IPackageManifestStatus;
}

export interface IClusterServiceVersionCRDResource {
  kind: string;
  name: string;
  version: string;
}

export interface IClusterServiceVersionCRD {
  description: string;
  displayName: string;
  kind: string;
  name: string;
  version: string;
  resources: IClusterServiceVersionCRDResource[];
  specDescriptors: Array<{
    description: string;
    displayName: string;
    path: string;
    "x-descriptors": string[];
  }>;
  statusDescriptors: Array<{
    description: string;
    displayName: string;
    path: string;
    "x-descriptors": string[];
  }>;
}

export interface IClusterServiceVersionSpec {
  apiservicedefinitions: any;
  customresourcedefinitions: {
    owned?: IClusterServiceVersionCRD[];
  };
  description: string;
  displayName: string;
  icon: Array<{
    base64data: string;
    mediatype: string;
  }>;
  install: any;
  installModes: [
    { supported: boolean; type: "OwnNamespace" },
    { supported: boolean; type: "SingleNamespace" },
    { supported: boolean; type: "MultiNamespace" },
    { supported: boolean; type: "AllNamespaces" },
  ];
  keywords: string[];
  labels: any;
  links: Array<{
    name: string;
    url: string;
  }>;
  maintainers: Array<{
    email: string;
    name: string;
  }>;
  maturity: string;
  provider: {
    name: string;
  };
  selector: any;
  version: string;
}

export interface IClusterServiceVersion extends IResource {
  spec: IClusterServiceVersionSpec;
}

export interface IPkgRepositoryFilter {
  jq: string;
  variables?: { [key: string]: string };
}

export interface IInstalledPackageState {
  isFetching: boolean;
  error?: FetchError | CreateError | UpgradeError | RollbackError | DeleteError;
  items: InstalledPackageDetail[];
  listOverview?: InstalledPackageSummary[];
  selected?: CustomInstalledPackageDetail;
  // TODO(agamez): add tests for this new state field
  selectedDetails?: AvailablePackageDetail;
  resourceRefs?: ResourceRef[];
}

export interface IStoreState {
  router: RouterState;
  apps: IInstalledPackageState;
  auth: IAuthState;
  packages: IPackageState;
  config: IConfigState;
  kube: IKubeState;
  repos: IPackageRepositoryState;
  clusters: IClustersState;
  operators: IOperatorsState;
}

interface IK8sResource {
  apiVersion: string;
  kind: string;
}

/** @see https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#lists-and-simple-kinds */
export interface IK8sList<I, M> extends IK8sResource {
  items: I[];
  metadata?: {
    resourceVersion?: string;
    selfLink?: string; // Not in docs, but seems to exist everywhere
  } & M;
}

export interface IRBACRole {
  apiGroup: string;
  namespace?: string;
  clusterWide?: boolean;
  resource: string;
  verbs: string[];
}

export interface IKubeItem<T> {
  isFetching: boolean;
  item?: T;
  error?: Error;
}

export interface IKind {
  apiVersion: string;
  plural: string;
  namespaced: boolean;
}

export interface IKubeState {
  items: { [s: string]: IKubeItem<IResource | IK8sList<IResource, {}>> };
  subscriptions: { [s: string]: Subscription };
  // TODO(minelson): Remove kinds and kindsError once the operator support is
  // removed from the dashboard or replaced with a plugin.
  kinds: { [kind: string]: IKind };
  kindsError?: Error;
}

export interface IBasicFormParam {
  path: string;
  type?: "string" | "number" | "integer" | "boolean" | "object" | "array" | "null" | "any";
  value?: any;
  title?: string;
  minimum?: number;
  maximum?: number;
  render?: string;
  description?: string;
  customComponent?: object;
  enum?: string[];
  hidden?:
    | {
        event: DeploymentEvent;
        path: string;
        value: string;
        conditions: Array<{
          event: DeploymentEvent;
          path: string;
          value: string;
        }>;
        operator: string;
      }
    | string;
  children?: IBasicFormParam[];
}
export interface IBasicFormSliderParam extends IBasicFormParam {
  sliderMin?: number;
  sliderMax?: number;
  sliderStep?: number;
  sliderUnit?: string;
}

export interface CustomInstalledPackageDetail extends InstalledPackageDetail {
  revision: number;
}

//  enum for the type of package repository storage
export enum RepositoryStorageTypes {
  PACKAGE_REPOSITORY_STORAGE_HELM = "helm",
  PACKAGE_REPOSITORY_STORAGE_OCI = "oci",
  PACKAGE_REPOSITORY_STORAGE_CARVEL_INLINE = "inline",
  PACKAGE_REPOSITORY_STORAGE_CARVEL_IMAGE = "image",
  PACKAGE_REPOSITORY_STORAGE_CARVEL_IMGPKGBUNDLE = "imgpkgBundle",
  PACKAGE_REPOSITORY_STORAGE_CARVEL_HTTP = "http",
  PACKAGE_REPOSITORY_STORAGE_CARVEL_GIT = "git",
}

export enum PluginNames {
  PACKAGES_HELM = "helm.packages",
  PACKAGES_FLUX = "fluxv2.packages",
  PACKAGES_KAPP = "kapp_controller.packages",
}

export interface IPkgRepoFormData {
  authMethod: PackageRepositoryAuth_PackageRepositoryAuthType;
  // kubeapps-managed secrets
  authHeader: string; // used if type == PACKAGE_REPOSITORY_AUTH_TYPE_BEARER || type == PACKAGE_REPOSITORY_AUTH_TYPE_AUTHORIZATION_HEADER
  basicAuth: UsernamePassword; // used if type == PACKAGE_REPOSITORY_AUTH_TYPE_BASIC_AUTH
  dockerRegCreds: DockerCredentials; // used if type == PACKAGE_REPOSITORY_AUTH_TYPE_DOCKER_CONFIG_JSON
  sshCreds: SshCredentials; // used if type == PACKAGE_REPOSITORY_AUTH_TYPE_SSH
  opaqueCreds: OpaqueCredentials; // used if type == PACKAGE_REPOSITORY_AUTH_TYPE_OPAQUE
  tlsCertKey: TlsCertKey; // used if type == PACKAGE_REPOSITORY_AUTH_TYPE_TLS
  // user-managed secrets
  secretAuthName: string;
  secretTLSName: string;
  // rest of the parameters
  customCA: string;
  description: string;
  interval: string;
  name: string;
  passCredentials: boolean;
  plugin: Plugin;
  skipTLS: boolean;
  type: string;
  url: string;
  // add more types if necessary
  customDetail?: Partial<
    HelmPackageRepositoryCustomDetail | KappControllerPackageRepositoryCustomDetail
  >;
  namespace: string;
  isNamespaceScoped: boolean;
}
