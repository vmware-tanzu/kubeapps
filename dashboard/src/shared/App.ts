import {
  AvailablePackageDetail,
  AvailablePackageReference,
  Context,
  CreateInstalledPackageRequest,
  InstalledPackageReference,
  ReconciliationOptions,
  VersionReference,
} from "gen/kubeappsapis/core/packages/v1alpha1/packages";
import * as url from "shared/url";
import { axiosWithAuth } from "./AxiosInstance";
import { KubeappsGrpcClient } from "./KubeappsGrpcClient";

export const KUBEOPS_ROOT_URL = "api/kubeops/v1";
export class App {
  private static client = () => new KubeappsGrpcClient().getPackagesServiceClientImpl();

  public static async GetInstalledPackageSummaries(
    cluster: string,
    namespace?: string,
    page?: number,
    size?: number,
  ) {
    return await this.client().GetInstalledPackageSummaries({
      context: { cluster: cluster, namespace: namespace },
      paginationOptions: { pageSize: size || 0, pageToken: page?.toString() || "0" },
    });
  }

  public static async GetInstalledPackageDetail(installedPackageRef?: InstalledPackageReference) {
    return await this.client().GetInstalledPackageDetail({
      installedPackageRef: installedPackageRef,
    });
  }

  public static async createInstalledPackage(
    targetContext: Context,
    name: string,
    availablePackageRef: AvailablePackageReference,
    pkgVersionReference: VersionReference,
    values?: string,
    reconciliationOptions?: ReconciliationOptions,
  ) {
    return await this.client().CreateInstalledPackage({
      name,
      values,
      targetContext,
      availablePackageRef,
      pkgVersionReference,
      reconciliationOptions,
    } as CreateInstalledPackageRequest);
  }

  public static async upgrade(
    installedPackageRef: InstalledPackageReference,
    chartNamespace: string,
    availablePackageDetail: AvailablePackageDetail,
    values?: string,
  ) {
    const endpoint = url.kubeops.releases.get(
      installedPackageRef.context?.cluster ?? "",
      installedPackageRef.context?.namespace ?? "",
      installedPackageRef.identifier,
    );
    const { data } = await axiosWithAuth.put(endpoint, {
      appRepositoryResourceName:
        availablePackageDetail.availablePackageRef?.identifier.split("/")[0],
      appRepositoryResourceNamespace: chartNamespace,
      chartName: decodeURIComponent(availablePackageDetail.name),
      releaseName: installedPackageRef.identifier,
      values,
      version: availablePackageDetail.version?.pkgVersion,
    });
    return data;
  }

  public static async rollback(installedPackageRef: InstalledPackageReference, revision: number) {
    const endpoint = url.kubeops.releases.get(
      installedPackageRef.context?.cluster ?? "",
      installedPackageRef.context?.namespace ?? "",
      installedPackageRef.identifier,
    );
    const { data } = await axiosWithAuth.put(
      endpoint,
      {},
      {
        params: {
          action: "rollback",
          revision,
        },
      },
    );
    return data;
  }

  public static async delete(installedPackageRef: InstalledPackageReference, purge: boolean) {
    let endpoint = url.kubeops.releases.get(
      installedPackageRef.context?.cluster ?? "",
      installedPackageRef.context?.namespace ?? "",
      installedPackageRef.identifier,
    );
    if (purge) {
      endpoint += "?purge=true";
    }
    const { data } = await axiosWithAuth.delete(endpoint);
    return data;
  }

  // TODO(agamez): remove it once we return the generated resources as part of the InstalledPackageDetail.
  public static async getRelease(installedPackageRef?: InstalledPackageReference) {
    const { data } = await axiosWithAuth.get<{ data: { manifest: any } }>(
      url.kubeops.releases.get(
        installedPackageRef?.context?.cluster ?? "",
        installedPackageRef?.context?.namespace ?? "",
        installedPackageRef?.identifier ?? "",
      ),
    );
    return data.data;
  }
}
