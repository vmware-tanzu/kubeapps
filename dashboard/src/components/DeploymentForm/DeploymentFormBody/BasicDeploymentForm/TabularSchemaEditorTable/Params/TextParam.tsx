// Copyright 2019-2022 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

import { CdsControlMessage } from "@cds/react/forms";
import { CdsInput } from "@cds/react/input";
import { CdsSelect } from "@cds/react/select";
import { CdsTextarea } from "@cds/react/textarea";
import Column from "components/js/Column";
import Row from "components/js/Row";
import { isEmpty } from "lodash";
import { useState } from "react";
import { validateValuesSchema } from "shared/schema";
import { IAjvValidateResult, IBasicFormParam } from "shared/types";
import { basicFormsDebounceTime, getStringValue, getValueFromString } from "shared/utils";

export interface ITextParamProps {
  id: string;
  label: string;
  inputType?: "text" | "textarea" | "password" | string;
  param: IBasicFormParam;
  handleBasicFormParamChange: (
    param: IBasicFormParam,
  ) => (e: React.FormEvent<HTMLInputElement | HTMLTextAreaElement | HTMLSelectElement>) => void;
}

export default function TextParam(props: ITextParamProps) {
  const { id, label, inputType, param, handleBasicFormParamChange } = props;

  const [validated, setValidated] = useState<IAjvValidateResult>();
  const [currentValue, setCurrentValue] = useState(getStringValue(param.currentValue));
  const [isValueModified, setIsValueModified] = useState(false);
  const [timeout, setThisTimeout] = useState({} as NodeJS.Timeout);

  const onChange = (
    e: React.FormEvent<HTMLInputElement | HTMLTextAreaElement | HTMLSelectElement>,
  ) => {
    // update the current value
    setCurrentValue(e.currentTarget.value);
    setIsValueModified(
      getStringValue(e.currentTarget.value) !== getStringValue(param.currentValue),
    );

    // twofold validation: using the json schema (with ajv) and the html5 validation
    setValidated(validateValuesSchema(e.currentTarget.value, param.schema));
    e.currentTarget?.reportValidity();

    // Gather changes before submitting
    clearTimeout(timeout);
    const func = handleBasicFormParamChange(param);
    // The reference to target get lost, so we need to keep a copy
    const targetCopy = {
      currentTarget: {
        value: getStringValue(e.currentTarget?.value),
        type: param.type === "object" ? param.type : e.currentTarget?.type,
      },
    } as React.FormEvent<HTMLInputElement | HTMLTextAreaElement | HTMLSelectElement>;
    setThisTimeout(setTimeout(() => func(targetCopy), basicFormsDebounceTime));
  };

  const unsavedMessage = isValueModified ? "Unsaved" : "";
  const isDiffCurrentVsDefault =
    getStringValue(param.currentValue, param.type) !==
    getStringValue(param.defaultValue, param.type);
  const isDiffCurrentVsDeployed =
    getStringValue(param.currentValue, param.type) !==
    getStringValue(param.defaultValue, param.type);
  const isModified =
    isValueModified ||
    (isDiffCurrentVsDefault && (!param.deployedValue || isDiffCurrentVsDeployed));

  const renderControlMsg = () =>
    !validated?.valid && !isEmpty(validated?.errors) ? (
      <>
        <CdsControlMessage status="error">
          {unsavedMessage}
          <br />
          {validated?.errors?.map(e => e.message).join(", ")}
        </CdsControlMessage>
        <br />
      </>
    ) : (
      <CdsControlMessage>{unsavedMessage}</CdsControlMessage>
    );

  const renderInput = () => {
    if (!isEmpty(param.enum)) {
      return (
        <>
          <CdsSelect layout="horizontal">
            <select
              required={param.isRequired}
              disabled={param.readOnly}
              aria-label={label}
              id={id}
              value={getStringValue(currentValue, param.type)}
              onChange={onChange}
            >
              <option disabled={true} key={""}>
                {""}
              </option>
              {param?.enum?.map((enumValue: any) => (
                <option value={getValueFromString(enumValue)} key={enumValue}>
                  {enumValue}
                </option>
              ))}
            </select>
            {renderControlMsg()}
          </CdsSelect>
        </>
      );
    } else if (param.type === "string") {
      switch (inputType) {
        case undefined:
        case "textarea":
          return (
            <CdsTextarea className={isModified ? "bolder" : ""}>
              <textarea
                required={param.isRequired}
                disabled={param.readOnly}
                maxLength={param.maxLength}
                minLength={param.minLength}
                aria-label={label}
                id={id}
                value={getStringValue(currentValue, param.type)}
                onChange={onChange}
              />
              {renderControlMsg()}
            </CdsTextarea>
          );
        default:
          return (
            <>
              <datalist id={`${id}-examples`}>
                {param?.examples?.map((example: any) => (
                  <option value={getValueFromString(example)} key={example}>
                    {example}
                  </option>
                ))}
              </datalist>
              <CdsInput className={isModified ? "bolder" : ""}>
                <input
                  required={param.isRequired}
                  disabled={param.readOnly}
                  maxLength={param.maxLength}
                  minLength={param.minLength}
                  pattern={param.pattern}
                  aria-label={label}
                  id={id}
                  list={param.examples ? `${id}-examples` : ""}
                  type={inputType}
                  value={getStringValue(currentValue, param.type)}
                  onChange={onChange}
                />
                {renderControlMsg()}
              </CdsInput>
            </>
          );
      }
      // is an object
    } else {
      return (
        <>
          <CdsTextarea className={isModified ? "bolder" : ""}>
            <textarea
              required={param.isRequired}
              disabled={param.readOnly}
              maxLength={param.maxLength}
              minLength={param.minLength}
              aria-label={label}
              id={id}
              value={getStringValue(currentValue)}
              onChange={onChange}
            />
            {renderControlMsg()}
          </CdsTextarea>
        </>
      );
    }
  };

  return (
    <Row>
      <Column span={10}>{renderInput()}</Column>
    </Row>
  );
}
