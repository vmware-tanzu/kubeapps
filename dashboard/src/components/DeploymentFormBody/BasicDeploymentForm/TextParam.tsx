// Copyright 2019-2022 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

import { isEmpty, isNumber } from "lodash";
import { useEffect, useState } from "react";
import { IBasicFormParam } from "shared/types";

export interface IStringParamProps {
  id: string;
  label: string;
  inputType?: string;
  param: IBasicFormParam;
  handleBasicFormParamChange: (
    param: IBasicFormParam,
  ) => (e: React.FormEvent<HTMLInputElement | HTMLTextAreaElement | HTMLSelectElement>) => void;
}

function TextParam({ id, param, label, inputType, handleBasicFormParamChange }: IStringParamProps) {
  const [value, setValue] = useState((param.value || "") as any);
  const [valueModified, setValueModified] = useState(false);
  const [timeout, setThisTimeout] = useState({} as NodeJS.Timeout);
  const onChange = (
    e: React.FormEvent<HTMLInputElement | HTMLTextAreaElement | HTMLSelectElement>,
  ) => {
    setValue(e.currentTarget.value);
    setValueModified(true);
    // Gather changes before submitting
    clearTimeout(timeout);
    const func = handleBasicFormParamChange(param);
    // The reference to target get lost, so we need to keep a copy
    const targetCopy = {
      currentTarget: {
        value: e.currentTarget.value,
        type: e.currentTarget.type,
      },
    } as React.FormEvent<HTMLInputElement | HTMLTextAreaElement | HTMLSelectElement>;
    setThisTimeout(setTimeout(() => func(targetCopy), 500));
  };

  useEffect(() => {
    if ((isNumber(param.value) || !isEmpty(param.value)) && !valueModified) {
      setValue(param.value);
    }
  }, [valueModified, param.value]);

  let input = (
    <input
      id={id}
      onChange={onChange}
      value={value}
      className="clr-input deployment-form-text-input"
      type={inputType ? inputType : "text"}
    />
  );
  if (inputType === "textarea") {
    input = <textarea id={id} onChange={onChange} value={value} />;
  } else if (param.enum != null && param.enum.length > 0) {
    input = (
      <select id={id} onChange={handleBasicFormParamChange(param)} value={param.value}>
        {param.enum.map(enumValue => (
          <option key={enumValue}>{enumValue}</option>
        ))}
      </select>
    );
  }
  return (
    <div>
      <label
        htmlFor={id}
        className="centered deployment-form-label deployment-form-label-text-param"
      >
        {label}
      </label>
      {input}
      {param.description && <span className="description">{param.description}</span>}
    </div>
  );
}

export default TextParam;
