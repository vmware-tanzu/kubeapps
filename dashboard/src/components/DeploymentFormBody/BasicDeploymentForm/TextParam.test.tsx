// Copyright 2019-2022 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

import { mount } from "enzyme";
import React from "react";
import { act } from "react-dom/test-utils";
import { IBasicFormParam } from "shared/types";
import TextParam from "./TextParam";

jest.useFakeTimers();

const stringParam = { path: "username", value: "user", type: "string" } as IBasicFormParam;
const stringProps = {
  id: "foo",
  label: "Username",
  param: stringParam,
  handleBasicFormParamChange: jest.fn().mockReturnValue(jest.fn()),
};

it("should render a string parameter with title and description", () => {
  const wrapper = mount(<TextParam {...stringProps} />);
  const input = wrapper.find("input");
  expect(input.prop("value")).toBe(stringProps.param.value);
  expect(wrapper).toMatchSnapshot();
});

it("should set the input type as number", () => {
  const wrapper = mount(<TextParam {...stringProps} inputType={"number"} />);
  const input = wrapper.find("input");
  expect(input.prop("type")).toBe("number");
});

it("should forward the proper value when using a string parameter", () => {
  const handler = jest.fn();
  const handleBasicFormParamChange = jest.fn().mockReturnValue(handler);
  const wrapper = mount(
    <TextParam {...stringProps} handleBasicFormParamChange={handleBasicFormParamChange} />,
  );
  const input = wrapper.find("input");

  const event = { currentTarget: { value: "" } } as React.FormEvent<HTMLInputElement>;
  act(() => {
    (input.prop("onChange") as any)(event);
  });
  wrapper.update();
  jest.runAllTimers();

  expect(handleBasicFormParamChange).toHaveBeenCalledWith({
    path: "username",
    type: "string",
    value: "user",
  });
  expect(handler).toHaveBeenCalledWith(event);
});

it("should set the input value as empty if a string parameter value is not defined", () => {
  const tparam = { path: "username", type: "string" } as IBasicFormParam;
  const tprops = {
    id: "foo",
    name: "username",
    label: "Username",
    param: tparam,
    handleBasicFormParamChange: jest.fn().mockReturnValue(jest.fn()),
  };
  const wrapper = mount(<TextParam {...tprops} />);
  const input = wrapper.find("input");
  expect(input.prop("value")).toBe("");
});

const textAreaParam = {
  path: "configuration",
  value: "First line\nSecond line",
  type: "string",
} as IBasicFormParam;
const textAreaProps = {
  id: "bar",
  label: "Configuration",
  param: textAreaParam,
  handleBasicFormParamChange: jest.fn().mockReturnValue(jest.fn()),
  inputType: "textarea",
};

it("should render a textArea parameter with title and description", () => {
  const wrapper = mount(<TextParam {...textAreaProps} />);
  const input = wrapper.find("textarea");
  expect(input.prop("value")).toBe(textAreaProps.param.value);
  expect(wrapper).toMatchSnapshot();
});

it("should forward the proper value when using a textArea parameter", () => {
  const handler = jest.fn();
  const handleBasicFormParamChange = jest.fn().mockReturnValue(handler);
  const wrapper = mount(
    <TextParam {...textAreaProps} handleBasicFormParamChange={handleBasicFormParamChange} />,
  );
  const input = wrapper.find("textarea");

  const event = { currentTarget: { value: "" } } as React.FormEvent<HTMLInputElement>;
  act(() => {
    (input.prop("onChange") as any)(event);
  });
  wrapper.update();
  jest.runAllTimers();

  expect(handleBasicFormParamChange).toHaveBeenCalledWith({
    path: "configuration",
    type: "string",
    value: "First line\nSecond line",
  });
  expect(handler).toHaveBeenCalledWith(event);
});

it("should set the input value as empty if a textArea param value is not defined", () => {
  const tparam = { path: "configuration", type: "string" } as IBasicFormParam;
  const tprops = {
    id: "foo",
    name: "configuration",
    label: "Configuration",
    param: tparam,
    handleBasicFormParamChange: jest.fn().mockReturnValue(jest.fn()),
    inputType: "textarea",
  };
  const wrapper = mount(<TextParam {...tprops} />);
  const input = wrapper.find("textarea");
  expect(input.prop("value")).toBe("");
});

it("should render a string parameter as select with option tags", () => {
  const tparam = {
    path: "databaseType",
    value: "postgresql",
    type: "string",
    enum: ["mariadb", "postgresql"],
  } as IBasicFormParam;
  const tprops = {
    id: "foo",
    name: "databaseType",
    label: "databaseType",
    param: tparam,
    handleBasicFormParamChange: jest.fn().mockReturnValue(jest.fn()),
  };
  const wrapper = mount(<TextParam {...tprops} />);
  const input = wrapper.find("select");

  expect(wrapper.find("select").prop("value")).toBe(tparam.value);
  if (tparam.enum != null) {
    const options = input.find("option");
    expect(options.length).toBe(tparam.enum.length);

    for (let i = 0; i < tparam.enum.length; i++) {
      const option = options.at(i);
      expect(option.text()).toBe(tparam.enum[i]);
    }
  }
});

it("should forward the proper value when using a select", () => {
  const tparam = {
    path: "databaseType",
    value: "postgresql",
    type: "string",
    enum: ["mariadb", "postgresql"],
  } as IBasicFormParam;
  const tprops = {
    id: "foo",
    name: "databaseType",
    label: "databaseType",
    param: tparam,
  };
  const handler = jest.fn();
  const handleBasicFormParamChange = jest.fn().mockReturnValue(handler);
  const wrapper = mount(
    <TextParam {...tprops} handleBasicFormParamChange={handleBasicFormParamChange} />,
  );
  const input = wrapper.find("select");

  const event = { currentTarget: {} } as React.FormEvent<HTMLSelectElement>;
  act(() => {
    (input.prop("onChange") as any)(event);
  });

  expect(handleBasicFormParamChange.mock.calls[0][0]).toEqual({
    path: "databaseType",
    type: "string",
    value: "postgresql",
    enum: ["mariadb", "postgresql"],
  });
  expect(handler.mock.calls[0][0]).toMatchObject(event);
});

it("a change in the param property should update the current value", () => {
  const wrapper = mount(<TextParam {...stringProps} param={{ ...stringParam, value: "" }} />);
  const input = wrapper.find("input");
  expect(input.prop("value")).toBe("");

  wrapper.setProps({
    param: {
      ...stringParam,
      value: "foo",
    },
  });
  wrapper.update();
  expect(wrapper.find("input").prop("value")).toBe("foo");
});

it("a change in a number param property should update the current value", () => {
  const numberParam = { path: "replicas", value: 0, type: "number" } as IBasicFormParam;
  const wrapper = mount(<TextParam {...stringProps} param={numberParam} />);
  const input = wrapper.find("input");
  expect(input.prop("value")).toBe(0);

  wrapper.setProps({
    param: {
      ...numberParam,
      value: 1,
    },
  });
  wrapper.update();
  expect(wrapper.find("input").prop("value")).toBe(1);
});
