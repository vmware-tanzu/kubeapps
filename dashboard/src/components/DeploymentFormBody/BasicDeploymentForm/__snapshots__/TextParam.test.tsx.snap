// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a string parameter with title and description 1`] = `
<TextParam
  handleBasicFormParamChange={[MockFunction]}
  id="foo"
  label="Username"
  param={
    Object {
      "path": "username",
      "type": "string",
      "value": "user",
    }
  }
>
  <div>
    <label
      className="centered deployment-form-label deployment-form-label-text-param"
      htmlFor="foo"
    >
      Username
    </label>
    <input
      className="clr-input deployment-form-text-input"
      id="foo"
      onChange={[Function]}
      type="text"
      value="user"
    />
  </div>
</TextParam>
`;

exports[`should render a textArea parameter with title and description 1`] = `
<TextParam
  handleBasicFormParamChange={[MockFunction]}
  id="bar"
  inputType="textarea"
  label="Configuration"
  param={
    Object {
      "path": "configuration",
      "type": "string",
      "value": "First line
Second line",
    }
  }
>
  <div>
    <label
      className="centered deployment-form-label deployment-form-label-text-param"
      htmlFor="bar"
    >
      Configuration
    </label>
    <textarea
      id="bar"
      onChange={[Function]}
      value="First line
Second line"
    />
  </div>
</TextParam>
`;
